cuda/matrix/csr_kernels.cu

```diff
@@ -429,7 +511,143 @@ __global__ __launch_bounds__(64) void classical_spmv(
+            kernel::abstract_merge_path_spmv<items_per_thread>
+                <<<grid, block, 0, 0>>>(
+                    static_cast<IndexType>(a->get_size()[0]),
+                    as_cuda_type(a->get_const_values()),
+                    a->get_const_col_idxs(),
+                    as_cuda_type(a->get_const_row_ptrs()),
+                    as_cuda_type(a->get_const_srow()), as_cuda_type(b_vals),
+                    b->get_stride(), as_cuda_type(c_vals), c->get_stride(),
+                    as_cuda_type(row_out.get_data()),
+                    as_cuda_type(val_out.get_data()));
+            kernel::abstract_reduce<<<1, spmv_block_size>>>(
+                grid_num, as_cuda_type(val_out.get_data()),
+                as_cuda_type(row_out.get_data()), as_cuda_type(c_vals),
+                c->get_stride());
+
+        } else if (alpha != nullptr && beta != nullptr) {
+            const auto b_vals = b->get_const_values() + column_id;
+            auto c_vals = c->get_values() + column_id;
+            kernel::abstract_merge_path_spmv<items_per_thread>
+                <<<grid, block, 0, 0>>>(
+                    static_cast<IndexType>(a->get_size()[0]),
+                    as_cuda_type(alpha->get_const_values()),
+                    as_cuda_type(a->get_const_values()),
+                    a->get_const_col_idxs(),
+                    as_cuda_type(a->get_const_row_ptrs()),
+                    as_cuda_type(a->get_const_srow()), as_cuda_type(b_vals),
+                    b->get_stride(), as_cuda_type(beta->get_const_values()),
+                    as_cuda_type(c_vals), c->get_stride(),
+                    as_cuda_type(row_out.get_data()),
+                    as_cuda_type(val_out.get_data()));
+            kernel::abstract_reduce<<<1, spmv_block_size>>>(
+                grid_num, as_cuda_type(val_out.get_data()),
+                as_cuda_type(row_out.get_data()),
+                as_cuda_type(alpha->get_const_values()), as_cuda_type(c_vals),
+                c->get_stride());
+        } else {
+            GKO_KERNEL_NOT_FOUND;
+        }
+    }
+}
+
+GKO_ENABLE_IMPLEMENTATION_SELECTION(select_merge_path_spmv, merge_path_spmv);
+
+
+template <typename IndexType>
+int compute_items_per_thread(std::shared_ptr<const CudaExecutor> exec)
+{
+    const int version = exec->get_major_version()
+                        << 4 + exec->get_minor_version();
+    // 128 threads/block the number of items per threads
+    // 3.0 3.5: 6
+    // 3.7: 14
+    // 5.0, 5.3, 6.0, 6.2: 8
+    // 5.2, 6.1, 7.0: 12
+    int num_item = 6;
+    switch (version) {
+    case 0x50:
+    case 0x53:
+    case 0x60:
+    case 0x62:
+        num_item = 8;
+        break;
+    case 0x52:
+    case 0x61:
+    case 0x70:
+        num_item = 12;
+        break;
+    case 0x37:
+        num_item = 14;
+    }
+    // The calculation is based on size(IndexType) = 4
+    constexpr int index_scale = sizeof(IndexType) / 4;
```

Comment:
The restriction is purely artificial, for now it makes sense to have `int32` and `int64` as index types so we have that by default, but there is no reason to restrain Ginkgo to only these. 

It looks like this code would indeed break with `int64`, but maybe I am wrong.

Reply:
Actually, it may work with `int64`, my bad. You get `items_per_thread = num_item / 2`. What could be worth considering is the following:

```C++
constexpr int index_scale = 4 / sizeof(IndexType);
int items_per_thread = num_item * index_scale;
```
If sizeof(IndexType) == 4, it works as before.
sizeof(IndexType) == 8, it works as before (index_scale is 0.5)
sizeof(IndexType) == 2, it will give twice as much items.

It looks like that was sort of the intended purpose of this code (less items with bigger index types) by using the `4 bytes` as reference. Then it should be possible to scale up when using a smaller index type.

Reply:
I like the idea. We could also rewrite it a bit, so we only use integer operations: 
```c++
// Items per thread is scaled so that each thread 4 bytes
int items_per_thread = (num_items * 4) / sizeof(IndexType);
```
The comment might not be ideal as I am not 100% sure what it does in the end.
It should lead to the same result as @tcojean proposed version.

The question is if the code that uses `compute_items_per_thread` works with all the output for different `IndexTypes`. I guess I will test that by hand for `uin16` and `uint64` and see what happens.

Reply:
I use shared_memory  in the `merge_path`
`block_item = spmv_block_size * items_per_thread
 =  spmv_block_size * num_item * 4 / sizeof(IndexType)`

`__shared__ IndexType shared_row_ptrs[block_items]`

It is used in two place.
1. it stores the row_pointer in the block, IndexType is fine
2. it stores the index and value for reduction, so it needs (sizeof(IndexType)+sizeof(ValueType))*spmv_block_size
```
IndexType *tmp_ind = shared_row_ptrs;
ValueType *tmp_val = reinterpret_cast<ValueType *>(shared_row_ptrs + spmv_block_size);
```

if it satisfies `(sizeof(IndexType)+sizeof(ValueType))*spmv_block_size<=block_item * sizeof(IndexType)`, it can work.

By the way, the `num_item` is decided to make occupancy 100%, so the value can be tuned in the future.

I will change the formula @thoasm mentioned and add the condition check.

Reply:
When I use `IndexType = int64`, it compiles, however, the test `ConjugateTransposeIsEquivalentToRef` fails with:
```
[ RUN      ] Csr.ConjugateTransposeIsEquivalentToRef
MapSMtoCores for SM 7.5is undefined. The default value of 64 Cores/SM is used.
unknown file: Failure
C++ exception with description "/home/thomas/projects/ginkgo_github/cuda/matrix/csr_kernels.cu:981: conj_transpose is not implemented" thrown in the test body.
[  FAILED  ] Csr.ConjugateTransposeIsEquivalentToRef (755 ms)
```
The other tests run through without any issue.
We should definitely fix that, because `int64` is supposed to be supported by us.


When trying to use `IndexType = int16` (by adding it to the macro `GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE`), I cannot compile it because of:
- for `cuda/preconditioner/jacobi_utils.hpp:138:34`, I get the error: `get_supported_storage_reductions [...] is used but never defined`
- for `cuda/matrix/csr_kernels.cu(231)`: `error: no instance of function template "gko::kernels::cuda::csr::kernel::get_warp_start_idx" matches the argument list`
- `cuda/components/cooperative_groups.cuh(349)`:  `error: static assertion failed with "Unable to shuffle sizes which are not 4-byte multiples"`

and so on.

It seems like we would need to put a lot of effort into getting `int16` as an `IndexType` to work. Some of the errors sound like an `int` was used instead of an `IndexType` (might also be a template parameter that is `int` and used in a function with `IndexType`).